File: nodejsmodules/fs.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
# This file is part of qmljs, the QML/JS language support plugin for KDevelop
# Copyright (c) 2014 Denis Steckelmacher <steckdenis@yahoo.fr>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of
# the License or (at your option) version 3 or any later version
# accepted by the membership of KDE e.V. (or its successor approved
# by the membership of KDE e.V.), which shall act as a proxy
# defined in Section 14 of version 3 of the license.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import sys
sys.path.append('../qmlplugins')

from jsgenerator import *
from common import *

# Print the license of the generated file (the same as the one of this file)
license()
basicTypes(globals())
require('stream')
require('buffer')
require('event')

_function = 'function(){}'
_object = 'new Object()'
_date = 'new Date()'
_buffer = 'new buffer.Buffer()'
_stats = 'new Stats'
_callback = ('callback', _function)
_fd = ('fd', _int)
_len = ('len', _int)
_path = ('path', _string)
_uid = ('uid', _int)
_gid = ('gid', _int)
_mode = ('mode', _int)
_offset = ('offset', _int)
_length = ('length', _int)
_position = ('position', _int)
_filename = ('filename', _string)
_options = ('options', _object)
_data = ('data', _string)

Module().members(
    F(_void, 'rename', ('oldPath', _string), ('newPath', _string), _callback),
    F(_int, 'renameSync', ('oldPath', _string), ('newPath', _string)),
    F(_void, 'ftruncate', _fd, _len, _callback),
    F(_int, 'ftruncateSync', _fd, _len),
    F(_void, 'truncate', _path, _len, _callback),
    F(_int, 'truncateSync', _path, _len),
    F(_void, 'chown', _path, _uid, _gid, _callback),
    F(_int, 'chownSync', _path, _uid, _gid),
    F(_void, 'fchown', _fd, _uid, _gid, _callback),
    F(_int, 'fchownSync', _fd, _uid, _gid),
    F(_void, 'lchown', _path, _uid, _gid, _callback),
    F(_int, 'lchownSync', _path, _uid, _gid),
    F(_void, 'chmod', _path, _mode, _callback),
    F(_int, 'chmodSync', _path, _mode),
    F(_void, 'fchmod', _fd, _mode, _callback),
    F(_int, 'fchmodSync', _fd, _mode),
    F(_void, 'lchmod', _path, _mode, _callback),
    F(_int, 'lchmodSync', _path, _mode),
    F(_void, 'stat', _path, _callback),
    F(_void, 'lstat', _path, _callback),
    F(_void, 'fstat', _fd, _callback),
    F(_stats, 'statSync', _path),
    F(_stats, 'lstatSync', _path),
    F(_stats, 'fstatSync', _fd),
    F(_void, 'link', ('srcpath', _string), ('dstpath', _string), _callback),
    F(_int, 'linkSync', ('srcpath', _string), ('dstpath', _string)),
    F(_void, 'symlink', ('srcpath', _string), ('dstpath', _string), ('type', _string), _callback),
    F(_int, 'symlinkSync', ('srcpath', _string), ('dstpath', _string), ('type', _string)),
    F(_void, 'readlink', _path, _callback),
    F(_int, 'readlinkSync', _path),
    F(_void, 'realpath', _path, ('cache', _object), _callback),
    F(_string, 'realpathSync', _path, ('cache', _object)),
    F(_void, 'unlink', _path, _callback),
    F(_int, 'unlinkSync', _path),
    F(_void, 'rmdir', _path, _callback),
    F(_int, 'rmdirSync', _path),
    F(_void, 'mkdir', _path, _mode, _callback),
    F(_int, 'mkdirSync', _path, _mode),
    F(_void, 'readdir', _path, _callback),
    F(_array, 'readdirSync', _path),
    F(_void, 'close', _fd, _callback),
    F(_int, 'closeSync', _fd),
    F(_void, 'open', _path, ('flags', _string), _mode, _callback),
    F(_int, 'openSync', _path, ('flags', _string), _mode),
    F(_void, 'utimes', _path, ('atime', _date), ('mtime', _date), _callback),
    F(_int, 'utimesSync', _path, ('atime', _date), ('mtime', _date)),
    F(_void, 'futimes', _fd, ('atime', _date), ('mtime', _date), _callback),
    F(_int, 'futimesSync', _fd, ('atime', _date), ('mtime', _date)),
    F(_void, 'fsync', _fd, _callback),
    F(_int, 'fsyncSync', _fd),
    F(_void, 'write', _fd, _buffer, _offset, _length, _position, _callback),
    F(_int, 'writeSync', _fd, _buffer, _offset, _length, _position),
    F(_void, 'read', _fd, _buffer, _offset, _length, _position, _callback),
    F(_int, 'readSync', _fd, _buffer, _offset, _length, _position),
    F(_void, 'readFile', _filename, _options, _callback),
    F(_int, 'readFileSync', _filename, _options),
    F(_void, 'writeFile', _filename, _data, _options, _callback),
    F(_int, 'writeFileSync', _filename, _data, _options),
    F(_void, 'appendFile', _filename, _data, _options, _callback),
    F(_int, 'appendFileSync', _filename, _data, _options),
    F(_void, 'watchFile', _filename, _options, _callback),
    F(_void, 'unwatchFile', _filename, _callback),
    F('new FSWatcher()', 'watch', _filename, _options, _callback),
    F(_void, 'exists', _path, _callback),
    F(_bool, 'existsSync', _path),
    Class('Stats').members(
        F(_bool, 'isFile'),
        F(_bool, 'isDirectory'),
        F(_bool, 'isBlockDevice'),
        F(_bool, 'isCharacterDevice'),
        F(_bool, 'isSymbolicLink'),
        F(_bool, 'isFIFO'),
        F(_bool, 'isSocket'),
        Var(_int, 'dev'),
        Var(_int, 'ino'),
        Var(_int, 'mode'),
        Var(_int, 'nlink'),
        Var(_int, 'uid'),
        Var(_int, 'gid'),
        Var(_int, 'rdev'),
        Var(_int, 'size'),
        Var(_int, 'blksize'),
        Var(_int, 'blocks'),
        Var(_date, 'atime'),
        Var(_date, 'mtime'),
        Var(_date, 'ctime')
    ),
    F('new ReadStream()', 'createReadStream', _path, _options),
    Class('ReadStream').prototype('stream.Readable'),
    F('new WriteStream()', 'createWriteStream', _path, _options),
    Class('WriteStream').prototype('stream.Writable'),
    Class('FSWatcher').prototype('event.EventEmitter').members(
        F(_void, 'close')
    )
).print()

File: nodejsmodules/fs.js
/*
 * This file is part of qmljs, the QML/JS language support plugin for KDevelop
 * Copyright (c) 2014 Denis Steckelmacher <steckdenis@yahoo.fr>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License or (at your option) version 3 or any later version
 * accepted by the membership of KDE e.V. (or its successor approved
 * by the membership of KDE e.V.), which shall act as a proxy
 * defined in Section 14 of version 3 of the license.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
    
var stream = require("stream");

var buffer = require("buffer");

var event = require("event");

exports.rename = function (oldPath, newPath, callback) { return ; };
exports.rename("", "", function(){});

exports.renameSync = function (oldPath, newPath) { return 1; };
exports.renameSync("", "");

exports.ftruncate = function (fd, len, callback) { return ; };
exports.ftruncate(1, 1, function(){});

exports.ftruncateSync = function (fd, len) { return 1; };
exports.ftruncateSync(1, 1);

exports.truncate = function (path, len, callback) { return ; };
exports.truncate("", 1, function(){});

exports.truncateSync = function (path, len) { return 1; };
exports.truncateSync("", 1);

exports.chown = function (path, uid, gid, callback) { return ; };
exports.chown("", 1, 1, function(){});

exports.chownSync = function (path, uid, gid) { return 1; };
exports.chownSync("", 1, 1);

exports.fchown = function (fd, uid, gid, callback) { return ; };
exports.fchown(1, 1, 1, function(){});

exports.fchownSync = function (fd, uid, gid) { return 1; };
exports.fchownSync(1, 1, 1);

exports.lchown = function (path, uid, gid, callback) { return ; };
exports.lchown("", 1, 1, function(){});

exports.lchownSync = function (path, uid, gid) { return 1; };
exports.lchownSync("", 1, 1);

exports.chmod = function (path, mode, callback) { return ; };
exports.chmod("", 1, function(){});

exports.chmodSync = function (path, mode) { return 1; };
exports.chmodSync("", 1);

exports.fchmod = function (fd, mode, callback) { return ; };
exports.fchmod(1, 1, function(){});

exports.fchmodSync = function (fd, mode) { return 1; };
exports.fchmodSync(1, 1);

exports.lchmod = function (path, mode, callback) { return ; };
exports.lchmod("", 1, function(){});

exports.lchmodSync = function (path, mode) { return 1; };
exports.lchmodSync("", 1);

exports.stat = function (path, callback) { return ; };
exports.stat("", function(){});

exports.lstat = function (path, callback) { return ; };
exports.lstat("", function(){});

exports.fstat = function (fd, callback) { return ; };
exports.fstat(1, function(){});

exports.statSync = function (path) { return new Stats; };
exports.statSync("");

exports.lstatSync = function (path) { return new Stats; };
exports.lstatSync("");

exports.fstatSync = function (fd) { return new Stats; };
exports.fstatSync(1);

exports.link = function (srcpath, dstpath, callback) { return ; };
exports.link("", "", function(){});

exports.linkSync = function (srcpath, dstpath) { return 1; };
exports.linkSync("", "");

exports.symlink = function (srcpath, dstpath, type, callback) { return ; };
exports.symlink("", "", "", function(){});

exports.symlinkSync = function (srcpath, dstpath, type) { return 1; };
exports.symlinkSync("", "", "");

exports.readlink = function (path, callback) { return ; };
exports.readlink("", function(){});

exports.readlinkSync = function (path) { return 1; };
exports.readlinkSync("");

exports.realpath = function (path, cache, callback) { return ; };
exports.realpath("", new Object(), function(){});

exports.realpathSync = function (path, cache) { return ""; };
exports.realpathSync("", new Object());

exports.unlink = function (path, callback) { return ; };
exports.unlink("", function(){});

exports.unlinkSync = function (path) { return 1; };
exports.unlinkSync("");

exports.rmdir = function (path, callback) { return ; };
exports.rmdir("", function(){});

exports.rmdirSync = function (path) { return 1; };
exports.rmdirSync("");

exports.mkdir = function (path, mode, callback) { return ; };
exports.mkdir("", 1, function(){});

exports.mkdirSync = function (path, mode) { return 1; };
exports.mkdirSync("", 1);

exports.readdir = function (path, callback) { return ; };
exports.readdir("", function(){});

exports.readdirSync = function (path) { return []; };
exports.readdirSync("");

exports.close = function (fd, callback) { return ; };
exports.close(1, function(){});

exports.closeSync = function (fd) { return 1; };
exports.closeSync(1);

exports.open = function (path, flags, mode, callback) { return ; };
exports.open("", "", 1, function(){});

exports.openSync = function (path, flags, mode) { return 1; };
exports.openSync("", "", 1);

exports.utimes = function (path, atime, mtime, callback) { return ; };
exports.utimes("", new Date(), new Date(), function(){});

exports.utimesSync = function (path, atime, mtime) { return 1; };
exports.utimesSync("", new Date(), new Date());

exports.futimes = function (fd, atime, mtime, callback) { return ; };
exports.futimes(1, new Date(), new Date(), function(){});

exports.futimesSync = function (fd, atime, mtime) { return 1; };
exports.futimesSync(1, new Date(), new Date());

exports.fsync = function (fd, callback) { return ; };
exports.fsync(1, function(){});

exports.fsyncSync = function (fd) { return 1; };
exports.fsyncSync(1);

exports.write = function (fd, n, offset, length, position, callback) { return ; };
exports.write(1, e, 1, 1, 1, function(){});

exports.writeSync = function (fd, n, offset, length, position) { return 1; };
exports.writeSync(1, e, 1, 1, 1);

exports.read = function (fd, n, offset, length, position, callback) { return ; };
exports.read(1, e, 1, 1, 1, function(){});

exports.readSync = function (fd, n, offset, length, position) { return 1; };
exports.readSync(1, e, 1, 1, 1);

exports.readFile = function (filename, options, callback) { return ; };
exports.readFile("", new Object(), function(){});

exports.readFileSync = function (filename, options) { return 1; };
exports.readFileSync("", new Object());

exports.writeFile = function (filename, data, options, callback) { return ; };
exports.writeFile("", "", new Object(), function(){});

exports.writeFileSync = function (filename, data, options) { return 1; };
exports.writeFileSync("", "", new Object());

exports.appendFile = function (filename, data, options, callback) { return ; };
exports.appendFile("", "", new Object(), function(){});

exports.appendFileSync = function (filename, data, options) { return 1; };
exports.appendFileSync("", "", new Object());

exports.watchFile = function (filename, options, callback) { return ; };
exports.watchFile("", new Object(), function(){});

exports.unwatchFile = function (filename, callback) { return ; };
exports.unwatchFile("", function(){});

exports.watch = function (filename, options, callback) { return new FSWatcher(); };
exports.watch("", new Object(), function(){});

exports.exists = function (path, callback) { return ; };
exports.exists("", function(){});

exports.existsSync = function (path) { return true; };
exports.existsSync("");

/*
 * exports.Stats
 */
function Stats() { return ; }
exports.Stats = Stats;

exports.Stats.prototype.isFile = function () { return true; };
exports.Stats.prototype.isFile();

exports.Stats.prototype.isDirectory = function () { return true; };
exports.Stats.prototype.isDirectory();

exports.Stats.prototype.isBlockDevice = function () { return true; };
exports.Stats.prototype.isBlockDevice();

exports.Stats.prototype.isCharacterDevice = function () { return true; };
exports.Stats.prototype.isCharacterDevice();

exports.Stats.prototype.isSymbolicLink = function () { return true; };
exports.Stats.prototype.isSymbolicLink();

exports.Stats.prototype.isFIFO = function () { return true; };
exports.Stats.prototype.isFIFO();

exports.Stats.prototype.isSocket = function () { return true; };
exports.Stats.prototype.isSocket();

exports.Stats.prototype.dev = 1;

exports.Stats.prototype.ino = 1;

exports.Stats.prototype.mode = 1;

exports.Stats.prototype.nlink = 1;

exports.Stats.prototype.uid = 1;

exports.Stats.prototype.gid = 1;

exports.Stats.prototype.rdev = 1;

exports.Stats.prototype.size = 1;

exports.Stats.prototype.blksize = 1;

exports.Stats.prototype.blocks = 1;

exports.Stats.prototype.atime = new Date();

exports.Stats.prototype.mtime = new Date();

exports.Stats.prototype.ctime = new Date();


exports.createReadStream = function (path, options) { return new ReadStream(); };
exports.createReadStream("", new Object());

/*
 * exports.ReadStream
 */
function ReadStream() { return ; }
exports.ReadStream = ReadStream;
exports.ReadStream.prototype = stream.Readable;


exports.createWriteStream = function (path, options) { return new WriteStream(); };
exports.createWriteStream("", new Object());

/*
 * exports.WriteStream
 */
function WriteStream() { return ; }
exports.WriteStream = WriteStream;
exports.WriteStream.prototype = stream.Writable;


/*
 * exports.FSWatcher
 */
function FSWatcher() { return ; }
exports.FSWatcher = FSWatcher;
exports.FSWatcher.prototype = event.EventEmitter;

exports.FSWatcher.prototype.close = function () { return ; };
exports.FSWatcher.prototype.close();


